API: Fix tags update on updating a note

jyuvaraj03 opened the pull request:
<!--

Please prefix the title with the platform you are targetting:

- "Desktop" for the Windows/macOS/Linux app (Electron app)
- "Mobile" for the mobile app (or "Android" / "iOS" if the pull request only applies to one of the mobile platforms)
- "CLI" for the CLI app

If it's not related to any platform (such as a translation, change to the documentation, etc.), simply don't add a platform.

For example: "Desktop: Added new setting to change font", or "Mobile: Fixed config screen error"

PLEASE READ THE GUIDE FIRST: https://github.com/laurent22/joplin/blob/master/CONTRIBUTING.md

-->
Fixes #941 

**Before:**
![before-2020-03-04-002019](https://user-images.githubusercontent.com/29891001/75809479-d6f08100-5dae-11ea-9d50-a2f9befbe3b5.gif)

**After:**
![after-2020-03-04-000934](https://user-images.githubusercontent.com/29891001/75808388-04d4c600-5dad-11ea-865e-66af171f76ce.gif)

The issue as mentioned in #941 was that the tags were not updated. This was due to the fact that in `PUT` request method the tags were not explicitly set unlike how the `POST` request method was handled in the `action_notes` function. 

The fix I have given makes the necessary changes. I have also written unit tests for the code. Looking forward to any comments or suggestions!

jyuvaraj03 commented:
@PackElend label me please

mic704b commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -451,6 +451,23 @@ class Api {
+
+			if (!note) throw new ErrorNotFound();
+
+			let newNote = Object.assign({}, note, request.bodyJson(this.readonlyProperties('PUT')));
```
I'm not familiar with this interface so please help me understand. Is the `PUT` command intended to update an existing note?  Is this really a new note, or an updated one?

jyuvaraj03 replied:
Yes, the `PUT` command is intended to update an existing note. Perhaps, the variable name `newNote` was a bit confusing? I will change it to `updatedNote`.

mic704b commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -451,6 +451,23 @@ class Api {
+			if (!note) throw new ErrorNotFound();
+
+			let newNote = Object.assign({}, note, request.bodyJson(this.readonlyProperties('PUT')));
+			newNote = await Note.save(newNote, { userSideValidation: true });
```
What is the effect of user-side validation here?

jyuvaraj03 replied:
The code I have written here was actually refactored from the `defaultAction_` function which was previously called in the case of `PUT` methods. 

To be honest, I do not know exactly know what user-side validation does here. But since it was already present in the code that was executed (before this fix), I decided to leave it untampered in order not to break any functionality.

mic704b commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +326,28 @@ describe('services_rest_Api', function() {
+		const tag3 = await Tag.save({ title: 'mon étiquette 3' });
+		const note = await Note.save({
+			title: 'ma note un',
+			tags: `${tag1.title},${tag2.title}`,
```
Does this add the tags correclty?  Perhaps you could start the testing by getting the note tags to check the note is set up as you intend?

jyuvaraj03 replied:
Yes, it does add the tags in the note (checked in the desktop app). But I haven't checked the note in the unit test. I will rewrite the unit test to make sure it tests the notes rather than the response.

mic704b replied:
Since you have the response checks, they are still useful.  So just add the new checks, so it checks both.

mic704b commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +326,28 @@ describe('services_rest_Api', function() {
+		}));
+		expect(response3.tags === `${tag1.title},${tag2.title},${tag3.title}`).toBe(true);
+	}));
 });
```
What happens if we accidently add a tag name that doesn't exist? (Is it possible to do via this interface?  If it is, then perhaps a test would be good.)

jyuvaraj03 replied:
If the user sets a tag name that doesn't exist via this interface, it creates the tags and then adds it to the note. I forgot to write a test case for that. I will fix the unit test.

jyuvaraj03 commented:
> Should you also check the actual tags on the note note after each `PUT` (using `Tag` and/or `Note` methods to directly interrogate the associated tags)? I'm not sure the test is covering the issue.

Yes, you were right. I should have checked the actual tags on the notes themselves. I have since rewritten the unit tests to do this.

jyuvaraj03 commented:
I have done all the requested changes. Please review :)

mic704b commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -451,6 +451,23 @@ class Api {
 
+		if (request.method === 'PUT') {
+			const note = await Note.load(id);
+			const requestBody = JSON.parse(request.body);
```
For consistency we should change the name of 'requestBody' to match what is used in the 'POST' code (just above).

Also there is no need to create this so early in the function, so please move it closer to where it is first used.

jyuvaraj03 replied:
> Also there is no need to create this so early in the function, so please move it closer to where it is first used.

I have made this change. But, can I know why it is important to not create it early in the function?

mic704b commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -451,6 +451,23 @@ class Api {
+
+			if (requestBody.tags) {
+				const tagTitles = requestBody.tags.split(',');
+				await Tag.setNoteTagsByTitles(updatedNote.id, tagTitles);
```
`updatedNote.id` can be just `id`

mic704b commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -451,6 +451,23 @@ class Api {
+			if (!note) throw new ErrorNotFound();
+
+			let updatedNote = Object.assign({}, note, request.bodyJson(this.readonlyProperties('PUT')));
+			updatedNote = await Note.save(updatedNote, { userSideValidation: true });
```
You mentioned you borrowed this from the default handler.  So rather than duplicate the code here, why not call the default handler to do this?

jyuvaraj03 replied:
Yeah, I didn't think about that! I'll implement it now.

jyuvaraj03 commented:
Completed all the requested changes. Please review!

laurent22 commented:
@mic704b, I've sent you an invite to be a collaborator on the project, so when you think this one is ready feel free to merge it.

Please use the syntax `Platform: Fixes|Resolves #xxx: Comment`. The comment will appear in the changelog so it should be something not too technical (indeed the title of this PR would be good). So for example `Api: Resolves #941: Fix tags update on updating a note`

jyuvaraj03 commented:
Sorry, I accidentally requested a review! Please review at your own pace.

mic704b commented:
> @mic704b, I've sent you an invite to be a collaborator on the project, so when you think this one is ready feel free to merge it.

Thank you @laurent22! I am honoured to be invited.

mic704b commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +327,33 @@ describe('services_rest_Api', function() {
+		expect(response2.tags === `${tag2.title},${newTag.title}`).toBe(true);
+		expect(tagIds2.includes(tag2.id)).toBe(true);
+		expect(tagIds2.includes(newTag.id)).toBe(true);
+	}));
```
Please add a test case for when tags are an empty string.  It is not expected, but we could receive it, so we should test we handle it properly.

mic704b commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +327,33 @@ describe('services_rest_Api', function() {
+		const tagIds1 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(response1.tags === `${tag1.title},${tag3.title}`).toBe(true);
+		expect(tagIds1.includes(tag1.id)).toBe(true);
+		expect(tagIds1.includes(tag3.id)).toBe(true);
```
We should also confirm what happened to tag2. Its absence can be confirmed by adding a check on the length of `tagIds1`

mic704b commented:
@jyuvaraj03 looks good, nearly there.

jyuvaraj03 commented:
> Please add a test case for when tags are an empty string. It is not expected, but we could receive it, so we should test we handle it properly.

Thanks for pointing it out. That was an excellent catch! I hadn't handled that case. When the tags in the request is empty, the tags associated with the note do not get deleted like intended. Instead, they stay the same. 

So I handled it by checking if the `tags` property is explicitly given as `''` during the API request. If so, all the existing tags are removed from the corresponding note. If the `tags` property is not mentioned during the request, no change is made to the tags in the note.

jyuvaraj03 commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +327,50 @@ describe('services_rest_Api', function() {
+		const tagIds4 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(response4.tags === '').toBe(true);
+		expect(tagIds4.length === 0).toBe(true);
+	}));
```
In this test case, the `tags` property is mentioned explicitly with a `''`. So, all the tags associated with the note are removed from the note.

jyuvaraj03 commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +327,50 @@ describe('services_rest_Api', function() {
+		const tagIds3 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(tagIds3.length === 2).toBe(true);
+		expect(tagIds3.includes(tag2.id)).toBe(true);
+		expect(tagIds3.includes(newTag.id)).toBe(true);
```
This test case checks the case when the `tags` property is **not** mentioned during the request. So, the tags associated with the note does not differ from the previous test case.

Also, I haven't checked `response3.tags` because it is `undefined` when `tags` property isn't mentioned during the API request. So, I skipped that check.

mic704b replied:
> because it is `undefined` when `tags` property isn't mentioned

That can be tested too!  If it wasn't returned as undefined, then it would be a problem.

mic704b commented on ReactNativeClient/lib/services/rest/Api.js:
```diff
@@ -451,6 +451,24 @@ class Api {
+			} else if (requestNote.tags) {
+				const tagTitles = requestNote.tags.split(',');
+				await Tag.setNoteTagsByTitles(id, tagTitles);
+			}
```
Can this be reduced to a single if statement?  I don't think you need two branches.

jyuvaraj03 replied:
Yes, I just realized two branches weren't necessary. I changed the code to use just a single `if`.

mic704b commented on CliClient/tests/services_rest_Api.js:
```diff
@@ -326,4 +327,50 @@ describe('services_rest_Api', function() {
+		const tagIds4 = await NoteTag.tagIdsByNoteId(note.id);
+		expect(response4.tags === '').toBe(true);
+		expect(tagIds4.length === 0).toBe(true);
+	}));
```
Generally, for future maintenance, it would be better to separate the test cases into 3 or 4 smaller test cases rather than one big one. This also helps avoid using the 1,2,3,4 variable names too.  Sometimes it is too inefficient, but by default it is preferred.

jyuvaraj03 replied:
I restructured all the test cases into separate ones. But, I'm not sure if I did it right. I have used lengthy test case descriptions for some test cases. Please review and let me know if I'm doing it right.

mic704b replied:
Looks good!

laurent22 commented:
> One more question: does the API documentation need update? The documentation says PUT on a note is to set note properties, but doesn't list tags as a property. Perhaps it should be added to the list, or the description changed to something like `Sets the properties and tags of ...`. What do you think?

That's a good point, the doc should be updated. To do so @jyuvaraj03 you can update command-apidoc.js, then you can verify the output from /CliClient using `npm start -- apidoc`.

In fact, ideally we should support all the special properties that the POST method supports, such as body_html, image_data_url, crop_rect, etc. for consistency. That could be done with a bit of refactoring, but it's not needed for this PR.

jyuvaraj03 commented:
> One more question: does the API documentation need update? The documentation says PUT on a note is to set note properties, but doesn't list tags as a property. Perhaps it should be added to the list, or the description changed to something like `Sets the properties and tags of ...`. What do you think?

You're right. So, I updated the doc. I added the `tags` property to the list of properties.

jyuvaraj03 commented:
> That's a good point, the doc should be updated. To do so @jyuvaraj03 you can update command-apidoc.js, then you can verify the output from /CliClient using npm start -- apidoc.

Yes, I did update the command-apidoc.js file. But, I have a small question. These were the lines of code I changed in that file:
```
tableFields.push({
	name: 'tags',
	type: Database.enumId('fieldType', 'text'),
	description: 'Comma-separated list of tags. eg. `tag1,tag2`.',
});
```

I just uncommented them in this commit. These lines already existed, but you had commented them out. Was there a reason you did that? Am I missing something?

<hr> 

> In fact, ideally we should support all the special properties that the POST method supports, such as body_html, image_data_url, crop_rect, etc. for consistency. That could be done with a bit of refactoring, but it's not needed for this PR.

I would like to work on implementing these (possibly in another PR). How should I go about it? Should I create an issue for this and then submit a PR?

laurent22 commented:
Ok I think it's ready to merge now, thanks @jyuvaraj03 and thanks @mic704b for reviewing!

mic704b commented:
> I think it's ready to merge now

Definitely.

Good work @jyuvaraj03
